copy/multiple.go

```diff
@@ -87,17 +89,43 @@ func (c *copier) copyMultipleImages(ctx context.Context, policyContext *signatur
 	}
 
 	// Copy each image, or just the ones we want to copy, in turn.
+	optionsInstances := options.Instances
 	instanceDigests := updatedList.Instances()
+	if len(options.InstancePlatforms) > 0 {
+		// Build a map for helping us avoid duplication of instance digests.
+		instanceDigestMap := make(map[digest.Digest]struct{})
+		for _, instanceDigest := range optionsInstances {
+			instanceDigestMap[instanceDigest] = struct{}{}
+		}
+		// Choose the best match for each platform we were asked to also copy.
+		for _, platform := range options.InstancePlatforms {
+			platformContext := types.SystemContext{
+				OSChoice:           platform.OS,
+				ArchitectureChoice: platform.Architecture,
+				VariantChoice:      platform.Variant,
+			}
+			instanceDigest, err := updatedList.ChooseInstance(&platformContext)
+			if err != nil {
+				return nil, fmt.Errorf("While choosing the instance for platform spec %q: %w", platform, err)
+			}
+			instanceDigestMap[instanceDigest] = struct{}{}
+		}
+		// Deduplicate the set of digests.
+		optionsInstances = nil
+		for instanceDigest := range instanceDigestMap {
+			optionsInstances = append(optionsInstances, instanceDigest)
+		}
+	}
 	imagesToCopy := len(instanceDigests)
 	if options.ImageListSelection == CopySpecificImages {
-		imagesToCopy = len(options.Instances)
+		imagesToCopy = len(optionsInstances)
 	}
 	c.Printf("Copying %d of %d images in list\n", imagesToCopy, len(instanceDigests))
 	updates := make([]manifest.ListUpdate, len(instanceDigests))
 	instancesCopied := 0
 	for i, instanceDigest := range instanceDigests {
 		if options.ImageListSelection == CopySpecificImages &&
-			!slices.Contains(options.Instances, instanceDigest) {
+			!slices.Contains(optionsInstances, instanceDigest) {
```

Comment:
This would ideally (at least in the asymptotic sense) be a hash table lookup, instead of an O(N) scan.

Currently we turn a list into a map, and then a map into a list; that second conversion could be just skipped, building a `digestsToCopy set.Set[digest.Digest]` from the combination of `options.Instances` and `options.InstancePlatforms`, and consuming it here directly. (We’d need a `set.Size()` added.)

Reply:
`len()` of the set's `Values()` will work until then.  I wouldn't complain about a `Size()` or `Len()`, though.

Reply:
This PR can add the `Size`/`Len`, it’s an internal subpackage we can change with no risk.

Reply:
You suggested `Size()` first, so I've added it as that.